include/libtorrent/packet_pool.hpp

```diff
@@ -84,26 +87,33 @@ namespace libtorrent
 		// deleter for std::unique_ptr
 		void operator()(packet* p) const
 		{
+			TORRENT_ASSERT(p != nullptr);
 			p->~packet();
 			std::free(p);
 		}
 	};
 
-	inline packet* create_packet(int size)
+	inline packet* create_packet(int const size)
 	{
 		packet* p = static_cast<packet*>(std::malloc(sizeof(packet) + size));
+		if (p == nullptr) aux::throw_ex<std::bad_alloc>();
 		new (p) packet();
+		p->allocated = aux::numeric_cast<std::uint16_t>(size);
```

Comment:
packet->allocated showed how many memory caller of acquire_packet wanted to allocate, but in reality packet occupied sizeof(packet) + slab.allocate_size (which may be great or equal to caller argument). 
i made assumption that only packets acquired from packet_pool will be released to same pool and less or equal logic was based on this fact. if you think that this "optimisation" will constrain somehow packets usage then replacing comparision with strict equal seems right.